### PR TITLE
mwindow: set limit on number of open files

### LINKED ISSUE #2758: GC mwindows on too many open files
We currently attempt to free unused map windows for packfiles when we go over a threshold. We however do not account for the number of open packfiles.

There's been at least two distinct reports of people running into issues when they have repositories which do network operations in small batches. These accumulate a lot of tiny packfiles and we end up having them all open at once, which can reach the system's limit for open file descriptors.

We should have a configurable limit on how many open file descriptors we want to have for packfiles and do a soft-GC similar to what we do for memory (address space).

### PR DESCRIPTION
**Pull request by josharian**
Doing many network operations can lead to lots of small packfiles.
The existing mwindow limit was only the total size of mmap'd files,
not on their number, so having lots of small packfiles could exhaust
the allowed number of open files, particularly on macOS, where
the default ulimit is very low.

Add a limit to the number of files as well as their total size,
and set the default limit low enough that even macOS users should not
hit it during normal use.

Fixes #2758

### CODE REVIEW

**Comment by josharian** [comment]
I'm not sure how to write a test for this.

Doing some manual testing (applied to master a little ways back, due to https://github.com/libgit2/libgit2/issues/5385), this helps with the situation in #2758, but not quite in the way I expected/hoped.

I still see O(#packfiles) open files using `lsof` with this patch applied. The difference this patch makes is that when the number of packfiles shrinks, so does the number of open files.

I don't understand enough about what's going on in the rest of the codebase that causes this combination of effects. I'd prefer that it strictly limited the number of open files. Hints and suggestions welcome!

For my purposes, this patch is at least enough to ameliorate the worst of #2758, because I can at least keep things under control by occasionally running `git gc`.

**Comment by lhchavez** [comment]
IIRC one way to reproduce this issue is to create >$OPEN_LIMIT packfiles where each packfile has a single commit object that points to a commit in another packfile. that way you never run into degenerate cases like needing a superlong chain of packfiles with deltas smeared among them (IIRC there is yet another codepath that limits the number of delta chains).

vanilla git and IIUC libgit2 (with this change applied) should both be able to handle that case correctly.

File: src/mwindow.c [hunk 1]
```diff
@@ -268,7 +272,8 @@ static git_mwindow *new_window(
 
 	ctl->mapped += (size_t)len;
 
-	while (git_mwindow__mapped_limit < ctl->mapped &&
+	while ((git_mwindow__mapped_limit < ctl->mapped ||
+	       git_mwindow__open_limit < ctl->open_windows) &&
```
Comment (pks-t) [inline src/mwindow.c]: Hum. I think this doesn't exactly work the way we need it to work.Thing is we track mwindows via a list ordered by recency of last usage. If we just call `git_mwindow_close_lru` until using less than `open_limit` files, then we will most likely close mwindows of other files, which doesn't help us at all in that case. Instead, we'd need to have some kind of heuristic that checks which file has the least frequently used mwindows and close all mwindows of that file only. Rinse and repeat until we are below `open_limit`.

Basically, instead of iterating over `git_mwindow`s, we need to iterate over `git_mwindow_file`s and determine which one of them to close.

Comment (josharian) [reply src/mwindow.c]: I see, I think.

From a quick look, that surgery might require more time than I really wanted to dedicate right now.

Do you think there is anyone who already knows this code that might be able to work on a proper fix? (If not, I may have to, since this issue is critical for me.)

Comment (lhchavez) [reply src/mwindow.c]: i may be able to help, since it's also affecting us. let me see if i can wrangle a few hours this weekend.

**Comment by pks-t** [comment]
On Thu, Feb 06, 2020 at 11:04:15AM -0800, lhchavez wrote:
> lhchavez commented on this pull request.
> > @@ -268,7 +272,8 @@ static git_mwindow *new_window(
>  
>  	ctl->mapped += (size_t)len;
>  
> -	while (git_mwindow__mapped_limit < ctl->mapped &&
> +	while ((git_mwindow__mapped_limit < ctl->mapped ||
> +	       git_mwindow__open_limit < ctl->open_windows) &&
> 
> i may be able to help, since it's also affecting us. let me see if i can wrangle a few hours this weekend.

Let me know if you don't find the time, then I could take over.
Shouldn't take too long, at least if I have a proper testcase.

File: src/mwindow.c (hunk 1 above)
Comment (lhchavez) [reply src/mwindow.c]: I could wrangle a few hours! #5396

Comment (josharian) [reply src/mwindow.c]: Hooray! Thanks!! I've been under water fighting another bug, but I'll patch in your PR and give it a whirl soon.

**Comment by ethomson** [comment]
Fixed via #5396
